torch/distributed/pipelining/schedules.py

```diff
@@ -2070,6 +2141,294 @@ def backward_stage_index(step):
+            )
+
+        # 1. Create the pipeline_order (all ranks do this calculation)
+        # This will be used to keep track of the current state of the entire pipeline
+        # pipeline_order[rank] = [Action(computation_type, microbatch_index, stage_index), ...]
+        self.pipeline_order: Dict[int, List[Optional[_Action]]] = self._get_zbv_schedule(
+            len(stages),
+            self.pp_group_size,
+            n_microbatches,
+        )
+        self.pipeline_order = _add_bubbles_to_actions(
+            self.n_local_stages * self.pp_group_size,
+            self.pipeline_order,
+            self.pp_group_size,
+        )
+
+    def _get_zbv_schedule(
+        self,
+        num_model_chunks,
+        pipeline_parallel_size,
+        num_microbatches,
+    ):
+        pp_group_size = pipeline_parallel_size
+        mb_size = num_microbatches
+        num_local_stages = num_model_chunks
+        compute_schedules = {}
+
+        def get_compute_schedule(pipeline_parallel_size, num_microbatches):
+            n_node = 6 * pipeline_parallel_size * num_microbatches
+
+            def get_id(cat, chunk, rank, micro):
+                return (
+                    cat * 2 * pipeline_parallel_size * num_microbatches
+                    + chunk * pipeline_parallel_size * num_microbatches
+                    + rank * num_microbatches
+                    + micro
+                )
+
+            count = []
+            for i in range(pipeline_parallel_size):
+                count.append([0] * 6)
+            fbw_mem = [39, -7, -32]
+            max_mem = 39 * (pipeline_parallel_size * 2)
+            end_time = [-1] * n_node
+            cur_time = [0] * pipeline_parallel_size
+            mem = [0] * pipeline_parallel_size
+            rank_bubble = [0] * pipeline_parallel_size
+            pending_w: List[Deque[Tuple[int, int, int]]] = [
+                deque() for _ in range(pipeline_parallel_size)
+            ]
+            schedule: Dict[int, List[Tuple[int, _ComputationType, int]]] = {}
+            for i in range(pipeline_parallel_size):
+                schedule[i] = []
+            stage_str = ["    " * i for i in range(pipeline_parallel_size)]
+            approved_bubble = [-1] * pipeline_parallel_size
+            max_approved_bubble = max(approved_bubble)
+
+            def get_max_rank_bubble(rank=-1):
+                max_rank_bubble = 0
+                for bb in rank_bubble:
+                    max_rank_bubble = max(max_rank_bubble, bb)
+                if rank >= 0:
+                    max_rank_bubble = max(
+                        max_rank_bubble, max_approved_bubble - approved_bubble[rank]
+                    )
+                return max_rank_bubble
+
+            def put_w(rank):
+                assert len(pending_w[rank]) > 0
+                _, chunk_, _ = pending_w[rank].popleft()
+                put(2, chunk_, rank)
+
+            def put(cat, chunk, rank, assert_cnt=True):
```

Comment:
would help to have description of what 'cat' means, especially.  maybe chunk is intuitive but idk yet.